### PR TITLE
Add groovy-macro docs

### PR DESCRIPTION
**Pull request by mariogarcia**
**WIP**

This is a **Working In Progress** because:

**Failing Tests**

There are 2 tests that started failing once I added the `groovy-macro` dependencies in specs:

- `org.codehaus.groovy.classgen.asm.sc.bugs.Groovy7316Bug > testTypeCheckingBypassUsingExplicitTypeHint`

- `org.codehaus.groovy.classgen.asm.sc.bugs.Groovy6757Bug > testExplicitTypeHint`

**Questions**

- Apart from that, there're a lot of licenses added I don't know where they came from. Should I exclude them from the PR ?

**Feedback**

Of course, need to know if the docs are ok, or not. Even if the PR is rejected I'd be happy if at least part of them help to the macros documentation.

### CODE REVIEW

**Comment by paulk-asert** [comment]
Those licenses should be generated into the root directory not src/spec/assets. You should exclude them from your PR.

**Comment by mariogarcia** [comment]
Ok, I've excluded the licenses.

File: src/spec/test/metaprogramming/ASTMatcherFilteringTest.groovy [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+            .visit(new ConvertToOne(source))  // <1>
+    }
+}
+// end::jokingtransformation[]
+
+// tag::jokingexpressiontransformer[]
+class ConvertToOne extends ClassCodeExpressionTransformer {
+    SourceUnit sourceUnit
+
+    ConvertToOne(SourceUnit sourceUnit1) {
```
Comment (bsideup) [inline src/spec/test/metaprogramming/ASTMatcherFilteringTest.groovy]: Minor cosmetics: `sourceUnit1` -> `sourceUnit` ?

File: src/spec/test/metaprogramming/ASTMatcherFilteringTest.groovy [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        this.sourceUnit = sourceUnit1
+    }
+
+    @Override
+    Expression transform(Expression exp) {
```
Comment (bsideup) [inline src/spec/test/metaprogramming/ASTMatcherFilteringTest.groovy]: this is just.... brilliant :D awesome example! :+1:

File: src/spec/test/metaprogramming/ASTMatcherFilteringTest.groovy [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        Expression ref = macro { 1 + 1 }     // <1>
+
+        if (ASTMatcher.matches(ref, exp)) {  // <2>
```
Comment (bsideup) [inline src/spec/test/metaprogramming/ASTMatcherFilteringTest.groovy]: Now I feel maybe we should implement `exp.matches(macro { 1 + 1 })`, or even `exp.matchesMacro { 1 + 1 }`, feels a bit more groovier :)

File: src/spec/test/metaprogramming/ASTMatcherTestingTest.groovy [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+class TwiceASTTransformation extends AbstractASTTransformation {
+
+    static final String VAR_X = 'x'
+
+    @Override
+    void visit(ASTNode[] nodes, SourceUnit source) {
+        ClassNode classNode = (ClassNode) nodes[1]
+        MethodNode giveMeTwo = getTemplateClass(sumExpression)
+            .methods
+            .find { it.name == 'giveMeTwo' }
```
Comment (bsideup) [inline src/spec/test/metaprogramming/ASTMatcherTestingTest.groovy]: can be replaced with `getTemplateClass(sumExpression).getDeclaredMethods('giveMeTwo').first()`

File: src/spec/test/metaprogramming/ASTMatcherFilteringTest.groovy (hunk 3 above)
Comment (mariogarcia) [inline src/spec/test/metaprogramming/ASTMatcherFilteringTest.groovy]: True. But why not simply overloading `matches` to receive a `Closure` ?

`exp.matches { 1 + 1 }`

File: src/spec/test/metaprogramming/ASTMatcherFilteringTest.groovy (hunk 1 above)
Comment (mariogarcia) [inline src/spec/test/metaprogramming/ASTMatcherFilteringTest.groovy]: :+1:

File: src/spec/test/metaprogramming/ASTMatcherTestingTest.groovy (hunk 4 above)
Comment (mariogarcia) [inline src/spec/test/metaprogramming/ASTMatcherTestingTest.groovy]: :+1:

File: src/spec/test/metaprogramming/ASTMatcherFilteringTest.groovy (hunk 3 above)
Comment (mariogarcia) [inline src/spec/test/metaprogramming/ASTMatcherFilteringTest.groovy]: Anyway this could be added in future versions

Comment (bsideup) [inline src/spec/test/metaprogramming/ASTMatcherFilteringTest.groovy]: sure, nothing to do with the documentation, just wanted to mention it :)

**Comment by mariogarcia** [comment]
BTW: I'm aware that commits must be rebased before merging them to the main branch (All have WIP...). I'm keeping them to keep track of the suggestions.

File: src/spec/doc/core-metaprogramming.adoc [hunk 5]
```diff
@@ -2829,6 +2829,153 @@ to use the Groovy Console, in particular the AST browser tool, to gain knowledge
 resource for learning is the https://github.com/apache/groovy/tree/master/src/test/org/codehaus/groovy/ast/builder[AST Builder]
 test suite.
 
+==== Macros
+
+===== Introduction
+
+Until version 2.5.0, when developing AST transformations, developers should have a deep knowledge about how the AST
+(Abstract source tree) was built by the compiler in order to know how to add new expressions or statements during
```
Comment (paulk-asert) [inline src/spec/doc/core-metaprogramming.adoc]: I think 'Abstract Syntax Tree' is more common than _source_ tree.

File: src/spec/doc/core-metaprogramming.adoc [hunk 6]
```diff
@@ -2829,6 +2829,153 @@ to use the Groovy Console, in particular the AST browser tool, to gain knowledge
...
+
+Although the use of `org.codehaus.groovy.ast.tool.GeneralUtils` static methods could mitigate the burden of creating
+expressions and statements, it's still a low-level way of writing those AST nodes directly.
+We needed something to abstract us from writing the AST directly and that's exactly what Groovy macros were made for.
+They allow you to add code during compile time directly, without having translate the code you had in mind to the
+`org.codehaus.groovy.ast.*` node related classes.
+
+===== Statements and expressions
+
+Lets see an example, lets create a local AST transformation. `@AddMessageMethod`. When applied to a given class it
```
Comment (paulk-asert) [inline src/spec/doc/core-metaprogramming.adoc]: lets => let's (by 2 cases)

File: src/spec/doc/core-metaprogramming.adoc [hunk 7]
```diff
@@ -2829,6 +2829,153 @@ to use the Groovy Console, in particular the AST browser tool, to gain knowledge
...
+will add a new method called `getMessage` to that class. The method will return "42". The annotation it's pretty
```
Comment (paulk-asert) [inline src/spec/doc/core-metaprogramming.adoc]: it's => is

File: src/spec/doc/core-metaprogramming.adoc [hunk 8]
```diff
@@ -2829,6 +2829,153 @@ to use the Groovy Console, in particular the AST browser tool, to gain knowledge
...
+straight forward:
+
+[source,groovy]
+----
+include::{projectdir}/src/spec/test/metaprogramming/MacroStatementTest.groovy[tags=addmethodannotation,indent=0]
+----
+
+How would look like the AST transformation without the use of a macro:
```
Comment (paulk-asert) [inline src/spec/doc/core-metaprogramming.adoc]: How would look like the AST transformation => What would the AST transformation look like

File: src/spec/doc/core-metaprogramming.adoc [hunk 9]
```diff
@@ -2829,6 +2829,153 @@ to use the Groovy Console, in particular the AST browser tool, to gain knowledge
...
+----
+include::{projectdir}/src/spec/test/metaprogramming/MacroStatementTest.groovy[tags=basicWithMacro,indent=0]
+----
+
+<1> Much simpler. You wanted to add a return statement that returned "42" and that's exactly what you can read inside
+the `macro` utility method. Your plain code will be translated for you to a `org.codehaus.groovy.ast.stmt.ReturnStatement`
+<2> Adding the return statement to the new method
+<3> Adding the new code to the annotated class
+
+Although `macro` method is used in this example to create an **statement** the `macro` method could also be used to create
```
Comment (paulk-asert) [inline src/spec/doc/core-metaprogramming.adoc]: Although `macro` => Although the `macro`
create an => create a

File: src/spec/doc/core-metaprogramming.adoc [hunk 10]
```diff
@@ -2829,6 +2829,153 @@ to use the Groovy Console, in particular the AST browser tool, to gain knowledge
...
+**expressions** as well, it depends on which `macro` signature you use:
+
+- `macro(Closure)`: Create a given statement with the code inside the closure.
+- `macro(Boolean,Closure)`: if **true** wrap expressions inside the closure inside an statement, if **false** then return
+an expression
+- `macro(CompilePhase, Closure)`: Create a given statement with the code inside the closure in a specific compile phase
+- `macro(CompilePhase, Boolean, Closure)`: Create an statement or an expression (true == statement, false == expression)
+in a specific compilation phase.
+
+NOTE: All this signatures can be found at `org.codehaus.groovy.macro.runtime.MacroGroovyMethods`
```
Comment (paulk-asert) [inline src/spec/doc/core-metaprogramming.adoc]: this signatures => these signatures

File: src/spec/doc/core-metaprogramming.adoc [hunk 11]
```diff
@@ -2829,6 +2829,153 @@ to use the Groovy Console, in particular the AST browser tool, to gain knowledge
...
+
+Sometimes we could be only interested in creating a given expression, not the whole statement, in order to do that we
+should use any of the `macro` invocations with a boolean parameter:
+
+[source,groovy]
+----
+include::{projectdir}/src/spec/test/metaprogramming/MacroExpressionTest.groovy[tags=addgettwotransformation,indent=0]
+----
+
+<1> We're telling macro not to wrap the expression in any statement, we're only interested in the expression
```
Comment (paulk-asert) [inline src/spec/doc/core-metaprogramming.adoc]: in any statement => in a statement

File: src/spec/doc/core-metaprogramming.adoc [hunk 12]
```diff
@@ -2829,6 +2829,153 @@ to use the Groovy Console, in particular the AST browser tool, to gain knowledge
...
+<4> Adding the code to the new method
+<5> Adding the method to the class
+
+===== Variable substitution
+
+Macros are great but we can't create anything useful or reusable if our macros couldn't receive parameters or resolve
+surrounding variables.
+
+In the following example we're creating an AST transformation `@MD5` that when applied to a given String field will
+add a method returning the MD5 the value of that field.
```
Comment (paulk-asert) [inline src/spec/doc/core-metaprogramming.adoc]: the MD5 the value => the MD5 value

File: src/spec/doc/core-metaprogramming.adoc [hunk 13]
```diff
@@ -2829,6 +2829,153 @@ to use the Groovy Console, in particular the AST browser tool, to gain knowledge
...
+
+And the transformation:
+
+[source,groovy]
+----
+include::{projectdir}/src/spec/test/metaprogramming/MacroVariableSubstitutionTest.groovy[tags=md5transformation,indent=0]
+----
+
+<1> We need a reference to a variable expression
+<2> If using a class outside the standard packages we should whether add needed imports or use the qualified name. When
```
Comment (paulk-asert) [inline src/spec/doc/core-metaprogramming.adoc]: whether add => add any

File: src/spec/doc/core-metaprogramming.adoc [hunk 14]
```diff
@@ -2829,6 +2829,153 @@ to use the Groovy Console, in particular the AST browser tool, to gain knowledge
...
+using the qualified named of a given static method you need to make sure it's resolved in the proper compile phase. In
+this particular case we're instructing the macro to resolve it at SEMANTIC_ANALYSIS, which is the first compile phase
```
Comment (paulk-asert) [inline src/spec/doc/core-metaprogramming.adoc]: at SEMANTIC_ANALYSIS => at the SEMANTIC_ANALYSIS phase

File: src/spec/doc/core-metaprogramming.adoc [hunk 15]
```diff
@@ -2829,6 +2829,153 @@ to use the Groovy Console, in particular the AST browser tool, to gain knowledge
...
+with type information.
+<3> In order to substitute any `expression` inside the macro we need to use the `$v` method. `$v` receives a closure as an
+argument, and the closure is only allowed to substitute expressions, meaning classes inheriting
+`org.codehaus.groovy.ast.expr.Expression`.
+
+===== MacroClass
+
+As we mentioned earlier the `macro` method is only capable of producing `statements` and `expressions`. But what if we
```
Comment (paulk-asert) [inline src/spec/doc/core-metaprogramming.adoc]: earlier the => earlier, the

File: src/spec/doc/core-metaprogramming.adoc [hunk 16]
```diff
@@ -2829,6 +2829,153 @@ to use the Groovy Console, in particular the AST browser tool, to gain knowledge
...
+want to produce other types of nodes, such a method, a field... ?
```
Comment (paulk-asert) [inline src/spec/doc/core-metaprogramming.adoc]: such a method, a field... ? => such as a method, a field, and so on?

File: src/spec/doc/core-metaprogramming.adoc [hunk 17]
```diff
@@ -2829,6 +2829,153 @@ to use the Groovy Console, in particular the AST browser tool, to gain knowledge
...
+
+`org.codehaus.groovy.macro.transform.MacroClass` can be used to create **classes** (ClassNode instances) in our
+transformations the same way we created statements and expressions with the `macro` method before.
+
+The next example is a local transformation `@Statistics`. When applied to a given class it will add two methods
```
Comment (paulk-asert) [inline src/spec/doc/core-metaprogramming.adoc]: class it => class, it

File: src/spec/doc/core-metaprogramming.adoc [hunk 18]
```diff
@@ -2829,6 +2829,153 @@ to use the Groovy Console, in particular the AST browser tool, to gain knowledge
...
+**getMethodCount()** and **getFieldCount()** which return how many methods and fields has the class respectively. Here
```
Comment (paulk-asert) [inline src/spec/doc/core-metaprogramming.adoc]: has the class => within the class

File: src/spec/doc/core-metaprogramming.adoc [hunk 19]
```diff
@@ -2829,6 +2829,153 @@ to use the Groovy Console, in particular the AST browser tool, to gain knowledge
...
+<6> Building the **getMethodCount()** method using reference's method count value expression
+<7> Building the **getFieldCount()** method using reference's field count value expression
+
+Basically we've created the **Statistics** class as a template to avoid writing low level AST API, then we
+copied methods created in the template class to their final destination.
+
+NOTE: Types inside the `MacroClass` implementation should be resolved inside, that's why I had to write
+`java.lang.Integer` instead of simply writing `Integer`.
+
+IMPORTANT: Notice that we'are using `@CompileDynamic`. That's because the way we use `MacroClass` is like we
```
Comment (paulk-asert) [inline src/spec/doc/core-metaprogramming.adoc]: we'are => we're

File: src/spec/doc/core-metaprogramming.adoc [hunk 20]
```diff
@@ -2868,6 +3015,90 @@ The difference is that when you use `assertScript`, the code in the `assertScrip
 unit test is executed*. That is to say that this time, the `Subject` class will be compiled with debugging active, and
 the breakpoint is going to be hit.
 
+===== ASTMatcher
+
+Sometimes you may want to make assertions over AST nodes whether to filter those nodes, or to make sure a given
```
Comment (paulk-asert) [inline src/spec/doc/core-metaprogramming.adoc]: nodes whether to filter those nodes => nodes; perhaps to filter the nodes

File: src/spec/doc/core-metaprogramming.adoc [hunk 21]
```diff
@@ -2829,6 +2829,153 @@ to use the Groovy Console, in particular the AST browser tool, to gain knowledge
...
+<3> Passing the reference class
+<4> Extracting reference class method count value expression
+<5> Extracting reference class field count value expression
+<6> Building the **getMethodCount()** method using reference's method count value expression
+<7> Building the **getFieldCount()** method using reference's field count value expression
+
+Basically we've created the **Statistics** class as a template to avoid writing low level AST API, then we
+copied methods created in the template class to their final destination.
+
+NOTE: Types inside the `MacroClass` implementation should be resolved inside, that's why I had to write
```
Comment (paulk-asert) [inline src/spec/doc/core-metaprogramming.adoc]: I => we

File: src/spec/doc/core-metaprogramming.adoc [hunk 22]
```diff
@@ -2868,6 +3015,90 @@ The difference is that when you use `assertScript`, the code in the `assertScrip
...
+
+Sometimes you may want to make assertions over AST nodes whether to filter those nodes, or to make sure a given
+transformation has built the expected AST node.
+
+**Filtering nodes**
+
+For instance if you would like to apply a given transformation only to a specific set of AST nodes, you could
+use **ASTMatcher** to filter these nodes. The following example shows how to transform a given expression to
+another. Using **ASTMatcher** it looks for a specific expression `1 + 1` and it transforms it to `3`. That's why
+I called it the `@Joking` example.
```
Comment (paulk-asert) [inline src/spec/doc/core-metaprogramming.adoc]: I => we

File: src/spec/test/metaprogramming/ASTMatcherFilteringTest.groovy [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+            .visit(new ConvertToOne(source))  // <1>
+    }
+}
+// end::jokingtransformation[]
+
+// tag::jokingexpressiontransformer[]
+class ConvertToOne extends ClassCodeExpressionTransformer {
+    SourceUnit sourceUnit
+
+    ConvertToOne(SourceUnit sourceUnit) {
```
Comment (paulk-asert) [inline src/spec/test/metaprogramming/ASTMatcherFilteringTest.groovy]: The name `ConvertToOne` is in some sense irrelevant but perhaps `ConvertToThree` or `ConvertOnePlusOneTo3` would be better?

File: src/spec/doc/core-metaprogramming.adoc [hunk 24]
```diff
@@ -2868,6 +3015,90 @@ The difference is that when you use `assertScript`, the code in the `assertScrip
...
+Then the transformation, that only applies an instance of `org.codehaus.groovy.ast.ClassCodeExpressionTransformer`
+to all the expressions within the method code block.
+
+[source,groovy]
+----
+include::{projectdir}/src/spec/test/metaprogramming/ASTMatcherFilteringTest.groovy[tags=jokingtransformation,indent=0]
+----
+<1> Get the method's code statement and apply the expression transformer
+
+And then is when the **ASTMatcher** is used to apply the transformation only to those expressions matching
```
Comment (paulk-asert) [inline src/spec/doc/core-metaprogramming.adoc]: And then => And this

File: src/spec/doc/core-metaprogramming.adoc [hunk 25]
```diff
@@ -2868,6 +3015,90 @@ The difference is that when you use `assertScript`, the code in the `assertScrip
...
+include::{projectdir}/src/spec/test/metaprogramming/ASTMatcherTestingTest.groovy[tags=twiceasttransformation,indent=0]
+----
+<1> Adding the method to the annotated class
+<2> Building a binary expression. The binary expression uses the same variable expression in both
+sides of the `+` token (check `varX` method at **org.codehaus.groovy.ast.tool.GeneralUtils**).
+<3> Builds a new **ClassNode** with a method called `giveMeTwo` which returns the result of an expression
+passed as parameter.
+
+Now instead of creating a test executing the transformation over a given sample code. I would like to check that
+the construction of the binary expression is donde properly:
```
Comment (paulk-asert) [inline src/spec/doc/core-metaprogramming.adoc]: donde => done (or perhaps reads better with 'performed')

File: src/spec/test/metaprogramming/ASTMatcherFilteringTest.groovy [hunk 26]
```diff
@@ -0,0 +1,103 @@
...
+import org.codehaus.groovy.transform.GroovyASTTransformation
+import org.codehaus.groovy.transform.GroovyASTTransformationClass
+
+import java.lang.annotation.ElementType
+import java.lang.annotation.Retention
+import java.lang.annotation.RetentionPolicy
+import java.lang.annotation.Target
+
+/**
+ * Created by dev on 6/30/16.
```
Comment (paulk-asert) [inline src/spec/test/metaprogramming/ASTMatcherFilteringTest.groovy]: I'd probably replace all of the `Created by dev on <_date_>` lines with a one line description of the test or just remove the comment.

**Comment by paulk-asert** [comment]
I added a bunch of mostly cosmetic suggestions. Looks really great overall. I also get the same test failures even though they seem unrelated at first glance. I'll look into that when I get the chance.

File: build.gradle [hunk 27]
```diff
@@ -238,6 +239,7 @@ dependencies {
 
     testCompile project(':groovy-ant')
     testCompile project(':groovy-test')
+    testCompile project(':groovy-macro')
```
Comment (paulk-asert) [inline build.gradle]: Adding this line alone is enough to cause the test failures. We obviously still need to determine why but at least we know your others changes are causing the problem.

File: src/spec/doc/core-metaprogramming.adoc (hunk 5 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: :+1:

File: src/spec/doc/core-metaprogramming.adoc [hunk 28]
```diff
@@ -2829,6 +2829,153 @@ to use the Groovy Console, in particular the AST browser tool, to gain knowledge
...
+===== Introduction
+
+Until version 2.5.0, when developing AST transformations, developers should have a deep knowledge about how the AST
+(Abstract source tree) was built by the compiler in order to know how to add new expressions or statements during
+compile time.
+
+Although the use of `org.codehaus.groovy.ast.tool.GeneralUtils` static methods could mitigate the burden of creating
+expressions and statements, it's still a low-level way of writing those AST nodes directly.
+We needed something to abstract us from writing the AST directly and that's exactly what Groovy macros were made for.
+They allow you to add code during compile time directly, without having translate the code you had in mind to the
```
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: :+1:

File: src/spec/doc/core-metaprogramming.adoc (hunk 7 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: :+1:

File: src/spec/doc/core-metaprogramming.adoc (hunk 8 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: :+1:

File: src/spec/doc/core-metaprogramming.adoc [hunk 29]
```diff
@@ -2829,6 +2829,153 @@ to use the Groovy Console, in particular the AST browser tool, to gain knowledge
...
+include::{projectdir}/src/spec/test/metaprogramming/MacroStatementTest.groovy[tags=addmethodtransformationwithoutmacro,indent=0]
+----
+
+<1> Create a return statement
+<2> Create a constant expression "42"
+<3> Adding the code to the new method
+<4> Adding the new method to the annotated class
+
+If you're not used to the AST API, that definitely doesn't look like the code you had in mind. Now look how the
+previous code looks like with the use of macros.
```
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: :+1:

File: src/spec/doc/core-metaprogramming.adoc (hunk 9 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: :+1:

File: src/spec/doc/core-metaprogramming.adoc (hunk 10 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: :+1:

File: src/spec/doc/core-metaprogramming.adoc [hunk 30]
```diff
@@ -2829,6 +2829,153 @@ to use the Groovy Console, in particular the AST browser tool, to gain knowledge
...
+should use any of the `macro` invocations with a boolean parameter:
+
+[source,groovy]
+----
+include::{projectdir}/src/spec/test/metaprogramming/MacroExpressionTest.groovy[tags=addgettwotransformation,indent=0]
+----
+
+<1> We're telling macro not to wrap the expression in any statement, we're only interested in the expression
+<2> Assigning the expression
+<3> Creating a `ReturnStatement` using a method from `GeneralUtils` and the expression returned
```
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: :+1:

File: src/spec/doc/core-metaprogramming.adoc (hunk 11 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: :+1:

File: src/spec/doc/core-metaprogramming.adoc (hunk 12 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: :+1:

File: src/spec/doc/core-metaprogramming.adoc (hunk 13 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: :+1:

File: src/spec/doc/core-metaprogramming.adoc (hunk 15 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: :+1:

File: src/spec/doc/core-metaprogramming.adoc (hunk 16 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: :+1:

File: src/spec/doc/core-metaprogramming.adoc (hunk 17 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: :+1:

File: src/spec/doc/core-metaprogramming.adoc (hunk 18 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: :+1:

File: src/spec/doc/core-metaprogramming.adoc (hunk 19 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: :+1:

File: src/spec/doc/core-metaprogramming.adoc (hunk 20 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: :+1:

File: src/spec/doc/core-metaprogramming.adoc (hunk 21 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: :+1:

File: src/spec/doc/core-metaprogramming.adoc (hunk 22 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: :+1:

File: src/spec/doc/core-metaprogramming.adoc (hunk 24 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: :+1:

File: src/spec/test/metaprogramming/ASTMatcherFilteringTest.groovy (hunk 26 above)
Comment (mariogarcia) [inline src/spec/test/metaprogramming/ASTMatcherFilteringTest.groovy]: :+1: I'll remove the comment

File: src/spec/test/metaprogramming/ASTMatcherFilteringTest.groovy (hunk 23 above)
Comment (mariogarcia) [inline src/spec/test/metaprogramming/ASTMatcherFilteringTest.groovy]: :+1:

File: src/spec/doc/core-metaprogramming.adoc (hunk 5 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: Changed!

File: src/spec/doc/core-metaprogramming.adoc (hunk 28 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: Changed!

File: src/spec/doc/core-metaprogramming.adoc (hunk 6 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: Changed!

File: src/spec/doc/core-metaprogramming.adoc (hunk 7 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: Changed!

File: src/spec/doc/core-metaprogramming.adoc (hunk 29 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: Changed!

File: src/spec/doc/core-metaprogramming.adoc (hunk 9 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: Changed!

File: src/spec/doc/core-metaprogramming.adoc (hunk 10 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: Changed!

File: src/spec/doc/core-metaprogramming.adoc (hunk 30 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: Changed!

File: src/spec/doc/core-metaprogramming.adoc (hunk 11 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: Changed!

File: src/spec/doc/core-metaprogramming.adoc (hunk 13 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: Changed!

File: src/spec/doc/core-metaprogramming.adoc (hunk 14 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: Changed!

File: src/spec/doc/core-metaprogramming.adoc (hunk 15 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: Changed!

File: src/spec/doc/core-metaprogramming.adoc (hunk 16 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: Changed!

File: src/spec/doc/core-metaprogramming.adoc (hunk 18 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: Changed!

File: src/spec/doc/core-metaprogramming.adoc (hunk 19 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: Changed!

File: src/spec/doc/core-metaprogramming.adoc (hunk 20 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: Changed!

File: src/spec/doc/core-metaprogramming.adoc (hunk 21 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: Changed!

File: src/spec/doc/core-metaprogramming.adoc (hunk 22 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: Changed!

File: src/spec/doc/core-metaprogramming.adoc (hunk 24 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: Changed!

File: src/spec/doc/core-metaprogramming.adoc (hunk 25 above)
Comment (mariogarcia) [inline src/spec/doc/core-metaprogramming.adoc]: Changed!

File: build.gradle (hunk 27 above)
Comment (bsideup) [inline build.gradle]: @paulk-asert let me know if you'll find anything!

File: src/spec/test/metaprogramming/MacroVariableSubstitutionTest.groovy [hunk 31]
```diff
@@ -0,0 +1,96 @@
...
+import org.codehaus.groovy.transform.GroovyASTTransformationClass
+
+import java.lang.annotation.ElementType
+import java.lang.annotation.Retention
+import java.lang.annotation.RetentionPolicy
+import java.lang.annotation.Target
+
+/**
+ *
+ */
```
Comment (paulk-asert) [inline src/spec/test/metaprogramming/MacroVariableSubstitutionTest.groovy]: can go too

File: src/spec/test/metaprogramming/ASTMatcherTestingTest.groovy [hunk 32]
```diff
@@ -0,0 +1,123 @@
...
+import java.lang.annotation.ElementType
+import java.lang.annotation.Retention
+import java.lang.annotation.RetentionPolicy
+import java.lang.annotation.Target
+
+import static org.codehaus.groovy.ast.ClassHelper.Integer_TYPE
+import static org.codehaus.groovy.ast.tools.GeneralUtils.*
+
+/**
+ * Created by dev on 6/30/16.
```
Comment (paulk-asert) [inline src/spec/test/metaprogramming/ASTMatcherTestingTest.groovy]: still 4 of these left

Comment (mariogarcia) [inline src/spec/test/metaprogramming/ASTMatcherTestingTest.groovy]: Sorry about that I should have review all of them again. All should be gone now :+1:

File: build.gradle (hunk 27 above)
Comment (jwagenleitner) [inline build.gradle]: PR #441 has been merged.

**Comment by paulk-asert** [comment]
Looking good. Any chance for a rebase?

**Comment by mariogarcia** [comment]
On it!

2016-10-13 9:32 GMT+02:00 Paul King <notifications@github.com>:

**Comment by mariogarcia** [comment]
Rebased and squashed!

**Comment by paulk-asert** [comment]
Merged. Thanks!!

**Comment by bsideup** [comment]
🎉  👍

**Comment by ilopmar** [comment]
@mariogarcia You are the man buddy! :tada:

**Comment by mariogarcia** [comment]
Yeahhh!!!!! :smiley: :tada: :confetti_ball: thanks @ilopmar I'm really happy to help version 2.5.0 to be a little bit closer. But of course this is teamwork, thanks to @bsideup and @paulk-asert as well.

**Pull request closed without merging**